### ISSUE TITLE
EmbeddingGemma: performance optimizations, test gaps, and upstream parity audit

### ISSUE THREAD

**Issue by SRjoeee** [findit-studio/mlxrs#332]
### Background

The EmbeddingGemma port (`embeddings/embeddinggemma/`, 4207 lines) is a faithful reproduction of the `google/embeddinggemma-300m` architecture (Gemma 3 text backbone as bidirectional encoder). Three expert audit teams verified 13 architecture components against upstream `mlx-embeddings/gemma3_text.py` and `mlx-lm/gemma3_text.py` — all match. Three divergences are all functionally equivalent or intentional improvements (Dense sanitize uses module-number instead of tensor shape, making it robust for quantized checkpoints where upstream's shape-based approach would break).

The shared infrastructure (`nn/quantized.rs`, `io/npy.rs`) is production-quality and already consumed by Whisper, SigLIP, and EmbeddingGemma.

---

#### P1. Fuse Q/K/V projections in encoder Attention [MEDIUM impact, MEDIUM difficulty]

**Location:** `backbone.rs:181-183`

```rust
let queries = self.q_proj.forward(x)?;
let keys = self.k_proj.forward(x)?;
let values = self.v_proj.forward(x)?;
```

Three separate linear projections over the same input `x`. For encoder inference (no KV cache, all three see the same input), a fused QKV matmul with a stacked `(q_out + 2*kv_out, hidden)` weight would:
- Reduce kernel launch overhead (1 matmul instead of 3)
- Improve memory access patterns (single read of `x`)
- Eliminate 2 extra graph nodes per layer (48 per forward for 24 layers)

With GQA (3 Q heads / 1 KV head), the fused weight construction needs care: Q gets `n_heads * head_dim` rows, K/V get `n_kv_heads * head_dim` rows each. The output is sliced after the single matmul.

The attention-pool head in SigLIP already uses this fused `in_proj` pattern — the same approach applies here.

#### P2. Weight transpose per-forward [LOW impact, MEDIUM difficulty]

**Location:** Every `linear()` call via `QuantizedLinear::forward` / `MaybeQuantizedLinear`

Each call creates a lazy `swapaxes` node before matmul. For 24 layers × (4 projections + 3 MLP) = ~168 transpose nodes per forward. MLX's backend optimizer may fuse these at eval time, but the graph-tracing overhead remains. Pre-transposing at load time would eliminate all of them.

#### P3. RoPE frequencies recomputed per-layer [LOW impact, LOW difficulty]

**Location:** `backbone.rs:149-157`, `lm/nn/rope.rs`

Each layer's `rope.apply()` invokes `mlx_fast_rope()` which computes cos/sin frequencies internally. For an encoder (offset=0, same sequence length every layer), the identical frequency tensor is recomputed 24 times. A `rope_with_freqs()` path exists where a precomputed frequency array could be cached and reused.

#### P4. Embedding scale scalar recomputed per-forward [LOW impact, TRIVIAL difficulty]

**Location:** `backbone.rs:493`

```rust
let scale = embedding_scale_like(self.hidden_size as f32, &h)?;
```

`sqrt(hidden_size)` is constructed as a fresh lazy array every forward. Cache as a field in `Gemma3Backbone` at load time.

#### P5. Mean pooling broadcasts mask to full hidden dimension [LOW impact, LOW difficulty]

**Location:** `pooling.rs:107-120`

Mask is broadcast from `(B, L, 1)` to `(B, L, H)` — full embedding dimension. MLX's optimizer likely handles this, but an alternative `sum(embeddings * mask, axis=1) / mask_sum` avoids the broadcast entirely.

---

#### T1. No E2E parity test [HIGH severity]

The test suite (1230 lines, ~30 tests) verifies shapes, invariants, and error paths but has **no numerical parity test** against the Python mlx-embeddings output. The tests.rs header explicitly states:

> "A google/embeddinggemma-300m checkpoint is not bundled, so there is no end-to-end cosine-parity oracle here."

This means the entire correctness claim rests on shape/invariant tests, not numerical agreement. A single E2e test with a known input → expected embedding (cosine ≥ 0.999 against Python output) would close this gap.

#### T2. Empty sequence causes division by zero in mean pooling [MEDIUM severity]

When `seq_len = 0` or all tokens are masked (`attention_mask` is all zeros), `mean_pooling` computes `sum(mask) = 0` and divides by zero. As a library, this should return a zero vector rather than NaN/Inf.

**Suggested fix:** Guard `mask_sum == 0` → return `Array::zeros` for that row.

#### T3. No all-weights-consumed assertion [MEDIUM severity]

`from_weights` succeeds without checking whether the weight map is empty afterward. An extra stray key (indicating a checkpoint layout mismatch) is silently ignored. Adding `if !weights.is_empty() { return Err(Error::UnexpectedKeys(...)) }` after building all components would catch this.

#### T4-T11. Missing test coverage [LOW severity]

| # | Missing test | Impact |
|---|-------------|--------|
| T4 | Isolated DenseHead forward test | Projection correctness unverified independently |
| T5 | Isolated mean_pooling test | Mask-aware averaging unverified with known values |
| T6 | Bidirectional attention test | No proof that position i attends to position i+1 |
| T7 | sliding_window > 0 validation | `sliding_window=0` would be accepted |
| T8 | Single-token (seq=1) input | Common degenerate case |
| T9 | Quantized vs dense numerical comparison | Only checks finite + L2-norm, not closeness |
| T10 | Duplicate keys in weight map | KeyCollision path untested |
| T11 | RoPE base frequency correctness | Global vs local layers getting the right theta |

---

### Part 3: Architecture parity summary

**13 MATCH:** RMSNorm, Attention projections, GQA, SDPA scale, MLP (gated GELU), TransformerBlock (sandwich-norm), RoPE, Embedding, DenseHead, Pooling, Task prefix, clip_residual, embedding_scale.

**3 DIVERGENCE (all improvements):**
1. q/k norm before vs after transpose — functionally equivalent (RMSNorm on head_dim)
2. Mask shape `(B,1,1,S)` vs `(B,1,S,S)` — mlxrs more efficient (broadcast equivalent)
3. Dense sanitize by module-number vs tensor shape — mlxrs more robust (upstream breaks on quantized checkpoints)

**1 MISSING:** `rope_scaling` config field — inconsequential for encoder.

**3 EXTRA:** Full quantization infrastructure, config validation with f32 narrowing checks, shape-pinned weight loading.

---

### Recommended implementation order

1. **T2 — Empty sequence guard** (quick correctness fix, prevents NaN in production)
2. **T3 — All-weights-consumed assertion** (quick safety fix)
3. **T1 — E2E parity test** (validates numerical correctness against reference)
4. **P1 — Fuse Q/K/V** (biggest performance win, standalone)
5. **P4 — Cache embedding_scale** (trivial, removes per-forward allocation)
6. **P3 — Cache RoPE frequencies** (low-hanging fruit)
7. **T4-T11 — Additional test coverage** (incremental)

**Comment by SRjoeee** [comment]
## Adversarial review — Part 1 perf, Part 2 test gaps, Part 3 upstream parity re-audit

Audited each part against the real code, with every verdict independently cross-checked, and **re-derived Part 3 parity from scratch** against the fetched upstream `gemma3_text.py` rather than trusting the "13 MATCH" claim. Findings below — **not self-fixing, leaving for the tech team.**

> Two constraints to disclose up front: (1) this environment has **no Metal compiler**, so I could not build/run the mlx-backed tests — all verification is static (code + fetched upstream); the key empirical claim (T2) is nonetheless settled by an *existing committed test*. (2) Verified premise: **no `mx.compile`/graph caching** anywhere — every forward re-traces and dispatches eagerly through mlx-c; `swapaxes`/`broadcast_to`/`reshape` are **metadata-only strided views**.

---

### Part 1 — Performance (all LOW/negligible; two are not real)

| Item | Problem real? | Verdict | Core issue |
|---|---|---|---|
| **P1 fuse QKV** | ✅ | **overstated, suboptimal** | GQA arithmetic is correct (`hidden=768, head_dim=256, n_heads=3, n_kv=1` → fused `(1280,768)`, slices `0:768/768:1024/1024:1280`). But the checkpoint ships **separate** q/k/v keys, so this is **concat-at-load**, not "slice the output". The real deployment is **4-bit quantized** (`group_size=64`), where `quantized_matmul` already does no transpose — so "fewer transposes / better memory access" largely doesn't apply, and you'd have to concat the packed **weight+scales+biases** triples (unaddressed). "-2 graph nodes/layer (48/fwd)" implies graph caching that doesn't exist. **The cited "mirrors SigLIP attn-pool in_proj" precedent is backwards** — `siglip2_naflex/vision.rs:190-194` *un-fuses* a stacked weight into 3 separate `linear()` calls. Severity MEDIUM → LOW; fuse at load + benchmark the quantized path, or decline. |
| **P2 weight transpose** | ❌ (quantized path) | **not real** | "168 transpose nodes via `QuantizedLinear::forward`" is path-confused: `quantized_matmul` (`quantized.rs:518-528`) uses an internal `transpose=true` flag — **zero** transpose graph nodes. The dense transpose (`quantized.rs:351`) is a free metadata-only view. EmbeddingGemma ships quantized. |
| **P3 cache RoPE freqs** | ❌ | **not real + correctness hazard** | The RoPE base **alternates per layer** (`backbone.rs:152-156`: global `rope_theta=1e6` vs local `rope_local_base_freq=1e4`) → there are **two** freq tensors, not one. "Cache **a** freq array" collapsed to one would **rotate 20 layers at the wrong base**. And the `freqs` param is only the tiny 128-element `base^(2i/dims)` seed (`rope.rs:269-274`), fused in-kernel — substituting it skips none of the dominant rotation, which runs every call regardless. Drop it. |
| **P4 cache embed scale** | ✅ | **trivial** | Real (`shared.rs:499` rebuilds a `(1,)` scalar/forward) but only graph-build overhead. Caveat: must keep the embedding-weight dtype, not assume f32. |
| **P5 mean-pool broadcast** | ✅ | **tiny** | `broadcast_to` is a free stride-0 view; the genuine waste is the full-size F32 mask cast (`pooling.rs:112`) + a redundant `hidden`× mask reduce (`:116`). The proposed fix leaves the dominant full-size product → sub-1%, unmeasurable. |

No benchmarks are cited for any item. Net: nothing here is worth shipping standalone.

---

### Part 2 — Correctness & test gaps (11 claims: 1 FALSE-bug, 3 already-covered, mixed rest)

**Meta-finding:** the audit looked only at `embeddinggemma/tests.rs` and **missed that pooling is shared infra** (`embeddings/pooling.rs` + `pooling/tests.rs`) with its own numerical + edge-case tests. That makes several "gaps" already-covered.

| Claim | Verdict | Evidence |
|---|---|---|
| **T2** div-by-zero in mean pooling | **FALSE** | `pooling.rs:116-119` already floors `sum_mask` at `1e-9` then divides → all-masked yields a finite **zero vector**, never NaN. Pinned by the committed test `mean_pooling_all_pad_row_uses_1e9_floor_finite_near_zero` (pooling/tests.rs:103). The proposed guard is redundant. **Close as not-a-bug.** |
| **T5** isolated mean_pooling | **FALSE** | `mean_pooling_hand_average_over_unmasked` (pooling/tests.rs:92) + others exist. |
| **T8** single-token seq=1 | **FALSE** | `single_token_sequence_all_strategies_return_that_token` (pooling/tests.rs:256). |
| **T1** no E2E parity | **PARTIAL** | Literal claim true (no real-checkpoint cosine oracle), but "shape/invariant tests only" overstates — closed-form numerical tests exist. **Downgrade HIGH→MEDIUM** and reframe: the locally-feasible gap is a **synthetic-weight numerical oracle** (pins RoPE-base-per-layer, the `sqrt(hidden)` scale, sandwich-norm order — things L2-norm/mask invariants can't catch). The real-checkpoint test should be hardware/network-gated, not committed-CI. |
| **T3** no all-weights-consumed check | **TRUE** | `from_weights` (mod.rs:163-185) has no `is_empty()` check. **But use the safe form:** emit a structured **warning** listing leftover keys, not a bare `return Err` — a hard reject is brittle against a future checkpoint with a tied `lm_head`/sidecar buffer. Add a regression test (none exists). |
| **T4** isolated DenseHead | TRUE (cosmetic) | No isolated test; covered transitively by every `encode_text`. |
| **T6** bidirectional attend i+1 | TRUE (low) | Structurally enforced (`build_additive_mask` emits no causal term); no direct assertion. |
| **T7** sliding_window>0 validation | TRUE (inert) | Not validated — **but the field is documented as unused** at inference (config.rs:113-115), so an invalid value can't affect output. Keep any check non-fatal. |
| **T9** quant-vs-dense closeness | **TRUE (most material)** | Quant tests assert only finite + L2-norm + shape (tests.rs:714-742, 855-908); none compares the **same** checkpoint dense-vs-quant, so a layout/dequant regression that still yields a finite unit vector passes silently. |
| **T10** KeyCollision path | TRUE (low) | Guarded (`insert_unique`, mod.rs:408) but only the *no-collision* case is tested (tests.rs:1063). |
| **T11** RoPE base wiring | TRUE (low) | `is_global_layer` predicate is tested (config/tests.rs:398), but not that the backbone wires global→`rope_theta`/local→`rope_local_base_freq`. |

Worth doing: **T3 (safe form), T1 (synthetic oracle), T9**. The rest are low/cosmetic; all additions are zero-risk test code except a hard-error form of T3/T7.

---

### Part 3 — Architecture parity (independently re-confirmed)

Three independent re-audits each re-derived parity from verbatim upstream (**mlx-embeddings** `gemma3_text.py`/`pooling.py`/`base.py` as the encoder reference + **mlx-lm** `gemma3_text.py`/`rope_utils.py` for the shared block), **without** relying on the prior claim. **All three confirm it: no real behavior-changing divergence on the forward path.** Verified MATCH: projection shapes, GQA (kv repeat internal to SDPA), SDPA scale (`query_pre_attn_scalar**-0.5`), RoPE base alternation (`traditional=False, scale=1.0`), gated `gelu_approx` MLP, the `sqrt(hidden)` embedding scale (**encoder form in weight dtype, not mlx-lm's bf16** — the "improvement" is the *correct* reference), `clip_residual` f16 saturation, the `1.0+weight` RMSNorm folded at load, bias-free no-activation DenseHead, L2 normalize, and config defaults vs the real `config.json` (`rope_scaling=null`).

The **3 claimed "improvement divergences" all hold up** under adversarial scrutiny:
1. **q/k RMSNorm before vs after transpose** — bit-identical (RMSNorm reduces only over the last axis = `head_dim`, which the `(0,2,1,3)` transpose never moves).
2. **mask `(B,1,1,S)` vs upstream `(B,1,S,S)`** — exactly equivalent (the upstream `mx.repeat` is a pure query-axis broadcast of a key-only pad mask; SDPA adds it with broadcasting) and avoids an O(S²) copy.
3. **Dense sanitize by module-number vs tensor-shape** — equal on every real (dense) checkpoint and **strictly more robust** on pre-quantized siblings, where upstream's shape heuristic misroutes packed `.weight`/`.scales`.

**One latent caveat worth a guard** (not a current bug): `config.rs` does not model `rope_scaling`, so the port silently assumes `rope_scaling=null` on global layers — true for every shipping checkpoint, but it would become a **real divergence** for any future Gemma3-encoder checkpoint that enables RoPE scaling. (The issue's "1 MISSING: rope_scaling — inconsequential" is right for today, but it's a latent risk, not strictly inconsequential.)

---

### Bottom line

- **Parity & correctness are clean** — the "13 MATCH + 3 benign" framing is substantiated, and the T2 "bug" does not exist.
- **All 5 perf items are LOW/negligible**; P2 and P3 aren't real (P3 is a correctness hazard as worded), and P1's headline rationale is overstated/backwards.
- **The genuinely worth-doing items are the tests, not the perf**: T9 (quant-vs-dense closeness), T1 (synthetic-weight numerical oracle), and T3 (stray-key detection, *warning* form). Recommend correcting the issue to credit the existing shared-pooling tests and close T2/T5/T8.
